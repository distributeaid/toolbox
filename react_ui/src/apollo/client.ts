import {
  ApolloClient,
  ApolloLink,
  fromPromise,
  HttpLink,
  InMemoryCache,
} from '@apollo/client'
import { CognitoUserSession } from 'amazon-cognito-identity-js'

import { Auth } from '../amplify'

const httpLink = new HttpLink({
  uri: '/api',
  credentials: 'same-origin',
  fetch,
})

const authLink = new ApolloLink((operation, forward) => {
  return fromPromise(
    Auth.currentSession()
      .then((session: CognitoUserSession) => {
        const token = session.getAccessToken().getJwtToken()

        operation.setContext({
          headers: {
            authorization: token ? `Bearer ${token}` : '',
          },
        })

        return operation
      })
      .catch(() => {
        // not signed in
        return operation
      })
  ).flatMap(forward)
})

export const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: ApolloLink.from([authLink.concat(httpLink)]),
})
