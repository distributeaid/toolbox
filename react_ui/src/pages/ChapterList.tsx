import React from 'react'
import { Trans, useTranslation } from 'react-i18next'

import { ContentContainer } from '../components/ContentContainer'
import { TextLink } from '../components/TextLink'
import { Group, useGetChapterListQuery } from '../generated/graphql'
import { NavLink } from 'react-router-dom'

export const ChapterItem: React.FC<{ chapter: Partial<Group> }> = ({
  chapter,
}) => {
  return (
    <li className="border-t border-gray-200">
      <NavLink to={`/${chapter.id}`}>
        <div className="py-4 hover:bg-gray-200">
          <div className="flex items-center justify-between">
            <div className="text-sm leading-5 font-medium text-black">
              {chapter.name}
            </div>

            <div className="ml-2 flex-shrink-0 flex">
              <span className="px-2 inline-flex text-xs leading-5 font-semibold">
                {/* {chapter.location.province}, {chapter.location.country.name} */}
              </span>
            </div>
          </div>
        </div>
      </NavLink>
    </li>
  )
}

export const ChapterList: React.FC = () => {
  const { t } = useTranslation()

  const { loading, error, data } = useGetChapterListQuery()

  if (loading || !data?.groups) {
    return null
  }

  if (error) {
    return <div>{error.toString()}</div>
  }

  const { groups: chapters } = data

  return (
    <ContentContainer>
      <div className="p-4 md:p-8">
        <div className="mb-4 md:mb-8">
          <h1 className="font-bold text-xl">{t('chapterList.title')}</h1>

          <p className="py-4">{t('chapterList.subtitle')}</p>

          <p>
            <Trans i18nKey="chapterList.instructions">
              If your area doesn't have a Masks For Docs Chapter yet, we are
              looking for volunteers to start new ones. Get in touch on{' '}
              <TextLink href="https://masksfordocs.slack.com">Slack</TextLink>
              to start one [Need better copy here].
            </Trans>
          </p>
        </div>

        <div className="grid grid-cols-1 gap-4">
          <ul>
            {chapters.map(
              (g) => g && g.id && <ChapterItem key={g.id} chapter={g} />
            )}
          </ul>
        </div>
      </div>
    </ContentContainer>
  )
}
