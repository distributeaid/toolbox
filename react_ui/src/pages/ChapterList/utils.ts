import { Group } from '../../generated/graphql'
import { usStateCodeToName } from '../../util/placeCodeToName'

export type CountryWithChapters = {
  countryCode: string
  chapters: Group[]
}

export const groupChaptersByCountry = (
  chapters: Group[]
): CountryWithChapters[] => {
  const chaptersByCountry: { [country: string]: Group[] } = {}

  for (const chapter of chapters) {
    const { countryCode } = chapter.location
    if (!chaptersByCountry[countryCode]) {
      chaptersByCountry[countryCode] = []
    }

    chaptersByCountry[countryCode].push(chapter)
  }

  return Object.keys(chaptersByCountry)
    .map((countryCode) => ({
      countryCode,
      chapters: chaptersByCountry[countryCode].sort((a, b) =>
        a.location.province
          .toLowerCase()
          .localeCompare(b.location.province.toLowerCase())
      ),
    }))
    .sort((a, b) =>
      a.countryCode.toLowerCase().localeCompare(b.countryCode.toLowerCase())
    )
}

export const provinceLongName = ({
  province,
  countryCode,
}: {
  province: string
  countryCode: string
}) => {
  if (countryCode === 'US') {
    return usStateCodeToName(province)
  }

  return province
}

export const provinceLocalName = (countryCode: string): string => {
  switch (countryCode) {
    case 'CA':
      return 'Province'
    default:
      return 'State'
  }
}
