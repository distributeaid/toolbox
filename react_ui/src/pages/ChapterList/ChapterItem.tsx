import React from 'react'
import { NavLink } from 'react-router-dom'

import { provinceLongName } from './utils'

export const ChapterItem: React.FC<{
  chapter: {
    name: string
    slug: string
    location: { province: string; countryCode: string }
  }
}> = ({ chapter }) => (
  <li className="border border-collapse border-gray-200 bg-white font-heading fill-black hover:border-blue-500 hover:shadow-xl hover:relative hover:z-10 hover:border hover:fill-pink">
    <NavLink
      to={`/${chapter.slug}`}
      className="p-6 flex flex-row text-black text-sm">
      <div className="leading-5 font-semibold w-4/12">
        {provinceLongName(chapter.location)}
      </div>
      <div className="leading-5 font-medium w-4/12">{chapter.name}</div>
      <div className="leading-5 w-4/12 flex justify-end">
        <svg
          width="17"
          height="17"
          viewBox="0 0 17 17"
          fill="none"
          xmlns="http://www.w3.org/2000/svg">
          <g clipPath="url(#clip0)">
            <path d="M1.82846 7.4855V9.4852L10.6065 9.4852L10.6065 13.4351L15.5562 8.48535L10.6065 3.5356L10.6065 7.4855L1.82846 7.4855Z" />
          </g>
          <defs>
            <clipPath id="clip0">
              <rect
                y="8.48535"
                width="12"
                height="12"
                transform="rotate(-45 0 8.48535)"
                fill="white"
              />
            </clipPath>
          </defs>
        </svg>
      </div>
    </NavLink>
  </li>
)
