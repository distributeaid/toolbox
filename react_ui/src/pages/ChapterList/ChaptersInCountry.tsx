import React from 'react'
import { Trans } from 'react-i18next'

import { P } from '../../components/P'
import { countryCodeToName } from '../../util/placeCodeToName'
import { ChapterItem } from './ChapterItem'
import { CountryWithChapters, provinceLocalName } from './utils'

export const ChaptersInCountry: React.FC<{
  t: (key: string) => string
  countryWithChapters: CountryWithChapters
}> = ({ countryWithChapters, t }) => (
  <div className="container p-6 pb-12 last:pb-6 md:mx-auto w-full max-w-4xl">
    <div className="grid grid-cols-1 gap-2">
      <h3 className="font-heading text-3xl font-semibold px-6">
        {countryCodeToName(countryWithChapters.countryCode)}
      </h3>

      <div className="flex px-6 font-heading font-semibold text-lg text-blue-600">
        <div className="w-4/12 pb-0">
          {provinceLocalName(countryWithChapters.countryCode)}
        </div>

        <div className="w-6/12 pb-0">{t('chapterList.chapterName')}</div>
      </div>

      <ul>
        {countryWithChapters.chapters.map((g) => (
          <ChapterItem key={g.id} chapter={g} />
        ))}
      </ul>
    </div>

    <P className="text-center mx-auto pt-6 w-2/3 sm:w-1/2">
      <Trans i18nKey="chapterList.cta">
        Donâ€™t see your local chapter? Reach out to{' '}
        <a
          href="https://share.hsforms.com/1ZEAK2ikxSFKqR_XR7J-x5Q4dh7j"
          className="underline text-blue-600"
          target="_blank"
          rel="noopener noreferrer">
          request supplies
        </a>
        ,{' '}
        <a
          href="https://share.hsforms.com/19adMT7wKRNSvxJSYJHOdGw4dh7j"
          className="underline text-blue-600"
          target="_blank"
          rel="noopener noreferrer">
          donate supplies
        </a>
        , or{' '}
        <a
          href="https://share.hsforms.com/1wEiqrs4DTQGBBQ9z2wBg8Q4dh7j"
          className="underline text-blue-600"
          target="_blank"
          rel="noopener noreferrer">
          volunteer
        </a>
        .
      </Trans>
    </P>
  </div>
)
