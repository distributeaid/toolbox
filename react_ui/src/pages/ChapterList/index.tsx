import React from 'react'
import { Trans, useTranslation } from 'react-i18next'

import { Background } from '../../components/Background'
import { ContentContainer } from '../../components/ContentContainer'
import { FAQFooter } from '../../components/FAQFooter'
import { MainContent } from '../../components/MainContent'
import { MainHeader } from '../../components/MainHeader'
import { PreHeader } from '../../components/PreHeader'
import { Group, Maybe, useGetChapterListQuery } from '../../generated/graphql'
import { ChaptersInCountry } from './ChaptersInCountry'
import { groupChaptersByCountry } from './utils'

export const ChapterList: React.FC = () => {
  const { t } = useTranslation()

  const { loading, error, data } = useGetChapterListQuery()

  if (loading || !data?.groups) {
    return null
  }

  if (error) {
    return <div>{error.toString()}</div>
  }

  const { groups: chapters } = data

  const chaptersByCountry = groupChaptersByCountry(
    (chapters as Maybe<Group>[]).filter(
      (chapter): chapter is Group => !!chapter
    )
  )

  const alphabetizedChaptersByCountry = chaptersByCountry.map((country) => {
    country.chapters.sort((a, b) =>
      a.name.toLowerCase().localeCompare(b.name.toLowerCase())
    )
    return country
  })

  return (
    <MainContent>
      <Background />
      <ContentContainer>
        <div className="md:w-7/12 pt-16 sm:pt-32">
          <MainHeader>{t('chapterList.title')}</MainHeader>

          <PreHeader>
            <Trans i18nKey="chapterList.subtitle">
              Connect locally to get what you need and give what you can. Find
              your local chapter below to{' '}
              <strong className="font-semibold">request supplies</strong>,{' '}
              <strong className="font-semibold">donate supplies</strong>, or{' '}
              <strong className="font-semibold">volunteer</strong>
            </Trans>
          </PreHeader>
        </div>
      </ContentContainer>

      <div className="border-gray-400 border-t w-full bg-lightBlue mt-20 py-20">
        {alphabetizedChaptersByCountry.map((cwc) => (
          <ChaptersInCountry
            key={cwc.countryCode}
            t={t}
            countryWithChapters={cwc}
          />
        ))}

        <ContentContainer>
          <FAQFooter />
        </ContentContainer>
      </div>
    </MainContent>
  )
}
