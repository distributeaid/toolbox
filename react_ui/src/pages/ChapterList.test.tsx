import { FetchResult } from '@apollo/client'
import { fireEvent, render, waitFor } from '@testing-library/react'
import { createMemoryHistory } from 'history'
import React, { Suspense } from 'react'
import { Router } from 'react-router-dom'

import {
  GetChapterListQuery,
  useGetChapterListQuery,
} from '../generated/graphql'
import { ChapterList } from './ChapterList'

jest.mock('react-i18next', () => ({
  useTranslation: () => {
    return { t: jest.fn().mockImplementation((v) => v) }
  },
  Trans: ({ children }: React.PropsWithChildren<{}>) => <>{children}</>,
}))

jest.mock('../generated/graphql', () => ({
  useGetChapterListQuery: jest.fn(),
  useGetChapterQuery: jest.fn(),
}))

it('loads the chapter list', async () => {
  ;(useGetChapterListQuery as jest.Mock).mockReturnValueOnce({
    loading: false,
    data: {
      groups: [
        {
          id: '1',
          name: 'Seattle',
          slug: 'wa-sea',
        },
        {
          id: '2',
          name: 'Oakland',
          slug: 'ca-oak',
        },
      ],
    } as FetchResult<GetChapterListQuery>,
  })

  const history = createMemoryHistory()
  const { findByText, getByText, container } = render(
    <Router history={history}>
      <Suspense fallback={null}>
        <ChapterList />
      </Suspense>
    </Router>
  )

  expect(await findByText('Seattle')).toBeVisible()
  expect(getByText('Oakland')).toBeVisible()

  expect(container).toMatchSnapshot()
  fireEvent.click(getByText('Oakland'))

  await waitFor(() =>
    expect(history.location).toMatchObject({ pathname: '/2' })
  )
})
