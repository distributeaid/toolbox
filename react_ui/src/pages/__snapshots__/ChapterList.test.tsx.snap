// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loads the chapter list 1`] = `
<div>
  <div
    class="container w-full max-w-6xl p-4 md:mx-auto bg-white overflow-hidden"
  >
    <div
      class="p-4 md:p-8"
    >
      <div
        class="mb-4 md:mb-8"
      >
        <h1
          class="font-bold text-xl"
        >
          chapterList.title
        </h1>
        <p
          class="py-4"
        >
          chapterList.subtitle
        </p>
        <p>
          If your area doesn't have a Masks For Docs Chapter yet, we are looking for volunteers to start new ones. Get in touch on
           
          <a
            class="underline text-blue-600 visited:text-purple-600"
            href="https://masksfordocs.slack.com"
            target=""
          >
            Slack
          </a>
          to start one [Need better copy here].
        </p>
      </div>
      <div
        class="grid grid-cols-1 gap-4"
      >
        <ul>
          <li
            class="border-t border-gray-200"
          >
            <a
              href="/wa-sea"
            >
              <div
                class="py-4 hover:bg-gray-200"
              >
                <div
                  class="flex items-center justify-between"
                >
                  <div
                    class="text-sm leading-5 font-medium text-black"
                  >
                    Seattle
                  </div>
                  <div
                    class="ml-2 flex-shrink-0 flex"
                  >
                    <span
                      class="px-2 inline-flex text-xs leading-5 font-semibold"
                    />
                  </div>
                </div>
              </div>
            </a>
          </li>
          <li
            class="border-t border-gray-200"
          >
            <a
              href="/ca-oak"
            >
              <div
                class="py-4 hover:bg-gray-200"
              >
                <div
                  class="flex items-center justify-between"
                >
                  <div
                    class="text-sm leading-5 font-medium text-black"
                  >
                    Oakland
                  </div>
                  <div
                    class="ml-2 flex-shrink-0 flex"
                  >
                    <span
                      class="px-2 inline-flex text-xs leading-5 font-semibold"
                    />
                  </div>
                </div>
              </div>
            </a>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
